include/seqan3/alignment/pairwise/detail/pairwise_alignment_algorithm_banded.hpp

```diff
@@ -0,0 +1,310 @@
+    template <indexed_sequence_pair_range indexed_sequence_pairs_t, typename callback_t>
+    //!\cond
+        requires std::invocable<callback_t, alignment_result_type>
+    //!\endcond
+    void operator()(indexed_sequence_pairs_t && indexed_sequence_pairs, callback_t && callback)
+    {
+        using result_value_t = typename alignment_result_value_type_accessor<alignment_result_type>::type;
+        using std::get;
+
+        for (auto && [sequence_pair, idx] : indexed_sequence_pairs)
+        {
+            result_value_t res{};
+            res.id = idx;
+
+            res.score = compute_matrix(get<0>(sequence_pair), get<1>(sequence_pair));
+            callback(alignment_result_type{res});
+        }
+    }
+    //!\}
+
+protected:
+    /*!\brief Checks wether the band given the sequence sizes is valid
+     *
+     * \param[in] sequence1_size The size of the first sequence.
+     * \param[in] sequence2_size The size of the second sequence.
+     *
+     * \throws seqan3::invalid_alignment_configuration if the band is invalid for the given sequence sizes and the
+     *         alignment configuration.
+     */
+    void check_valid_band_configuration(size_t const sequence1_size, size_t const sequence2_size) const
+    {
+        bool const upper_diagonal_ends_before_last_cell = (upper_diagonal + sequence2_size) < sequence1_size;
+        bool const lower_diagonal_ends_behind_last_cell = (-lower_diagonal + sequence1_size) < sequence2_size;
+
+        if (upper_diagonal_ends_before_last_cell || lower_diagonal_ends_behind_last_cell)
+            throw invalid_alignment_configuration{to_string("The selected band [",
+                                                            lower_diagonal,
+                                                            ":",
+                                                            upper_diagonal,
+                                                            "] does not cover the last cell.")};
+    }
+
+    /*!\brief Compute the actual banded alignment.
+     * \tparam sequence1_t The type of the first sequence; must model std::ranges::forward_range.
+     * \tparam sequence2_t The type of the second sequence; must model std::ranges::forward_range.
+     *
+     * \param[in] sequence1 The first sequence to compute the alignment for.
+     * \param[in] sequence2 The second sequence to compute the alignment for.
+     *
+     * \details
+     *
+     * In the banded alignment the iteration of the inner columns is split into two phases. The first phase
+     * reuses the unbanded column computation and assumes that the banded score matrix always starts at the beginning
+     * of the matrix. In the second phase the special interface
+     * seqan3::detail::pairwise_alignment_algorithm_banded::compute_band_column is used to compute the banded column.
+     * The current implementation assumes that the first cell of the current score matrix column is always the first
```

Comment:
This sounds like a future implementation will change this / relax this constraint. Is this your intention?

Reply:
I don't quite get the sentence maybe you can explain it to me in person :)

Reply:
Are you coming over 🍻? 
Not sure if this can be still be done with implementations like the waterman-eggert which requires the entire score matrix to be stored. So yes, this is for the moment the assumption we have. The point is that it is not so easy to explain. But let me try again.

The band starts to move downwards the matrix after the index given by upper_diagonal. So when upper_diagonal is 4. Then the first 5 columns (0, 1, 2, 3, 4) of the matrix will have an intersection with the band in the first row. These are cells that need to be intialised according to the respective recursion rule of the configured alignment. In the code we assume in the single column score matrix that the index `0` of that column is the one being initialised. When the band moves downwards starting at column 5 the first cell of the matrix is also the first cell of the band:

```
//  0 1 2 3 4 5 6
// |–|—|—|—|—|—|—|
//0|0|0|0|0|0| | |
//1|1|1|1|1|1|0| |
//2| |2|2|2|2|1|0|
//3| | |3|3|3|2|1|

//  0 1 2 3 4 5 6
// |–|—|—|—|—|—|—|
//0|4|3|2|1|0| | |
//1|5|4|3|2|1|0| |
//2| |5|4|3|2|1|0|
//3| | |5|4|3|2|1|
```
In the second example we would assume a different layout and would need to handle the computation differently. But with the upper scheme we can actually reuse the computation of the non-banded algorithm. Since the sequences determine the size of a band this is really nice, as very little things need to be changed/touched for the banded version.